File: packages/e2e/plugins/mock-cli/src/adapter.ts
// packages/e2e/plugins/mock-cli/src/adapter.ts
import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import type {
  Adapter,
  AdapterModel,
  AdapterSession,
  AgentConfig,
  Skill,
  SessionOptions,
  ToolCategories,
} from '@qlan-ro/mainframe-types';
import { ReplaySession } from './session';
import { createReplayState, type RecordedEvent } from './fixture';
import { listSkills as scanSkills, listAgents as scanAgents } from './skills';

function sanitizeKey(key: string): string {
  return key
    .replace(/[^a-zA-Z0-9._-]+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');
}

export class MockCliAdapter implements Adapter {
  id = 'mock-cli';
  name = 'Mock CLI';
  readonly capabilities = { planMode: true };
  private readonly indexByKey = new Map<string, number>();
  // Cache each live session's events keyed by its recorded sessionId (from onInit), so the extra
  // createSession that getMessagesFromDisk does (passing chatId) reuses the same events for
  // loadHistory instead of consuming the next fixture index.
  private readonly eventsBySessionId = new Map<string, RecordedEvent[]>();
  // Fallback for a history/resume read whose chatId isn't a recorded sessionId (Mainframe may set
  // claudeSessionId to a value the fixture didn't capture): reuse the most recent live session.
  private lastLiveEvents: RecordedEvent[] | null = null;

  async isInstalled(): Promise<boolean> {
    return true;
  }
  async getVersion(): Promise<string | null> {
    return '0.1.0';
  }
  async listModels(): Promise<AdapterModel[]> {
    return [
      {
        id: 'claude-haiku-4-5-20251001',
        label: 'Haiku 4.5',
        isDefault: true,
        // No capability fields — effort/features controls hide for this model.
      },
      {
        id: 'claude-sonnet-4-5-20251101',
        label: 'Sonnet 4.5',
        supportedEfforts: ['low', 'medium', 'high', 'max'],
        defaultEffort: 'medium',
        supportsFast: true,
      },
      {
        id: 'claude-opus-4-5-20251001',
        label: 'Opus 4.5',
        supportedEfforts: ['low', 'medium', 'high', 'xhigh', 'max'],
        defaultEffort: 'medium',
        supportsFast: true,
        supportsUltracode: true,
        supportsAdaptiveThinking: true,
      },
    ];
  }
  killAll(): void {}

  /**
   * Tool categorization so the display pipeline groups explore/progress/subagent tool calls the
   * same way it does for the real `claude` adapter (`prepareMessagesForClient` no-ops entirely
   * without this — see packages/core/src/messages/display-pipeline.ts). Mirrors Claude's
   * `explore`/`progress`/`subagent` sets exactly (packages/core/src/plugins/builtin/claude/
   * adapter.ts). `hidden` is deliberately left EMPTY rather than mirrored: Claude hides
   * AskUserQuestion/TodoWrite/EnterPlanMode raw tool cards, but tool-cards.spec.ts's already-
   * committed "AskUserQuestion display card" test relies on today's uncategorized (visible)
   * behavior — hiding it here would silently break that test. TaskCreate/TaskUpdate don't need to
   * be in `hidden` either: groupToolCallParts checks `progress` before `hidden`, so they're
   * captured into `_task_progress` regardless.
   */
  getToolCategories(): ToolCategories {
    return {
      explore: new Set(['Read', 'Glob', 'Grep', 'LS']),
      hidden: new Set(),
      progress: new Set(['TaskCreate', 'TaskUpdate']),
      subagent: new Set(['Task', 'Agent']),
    };
  }

  /** Project-scope only (no homedir scan) — see skills.ts. Lets a recording/e2e project that
   *  seeds `.claude/skills|agents` populate the Skills/Agents panels under mock-cli. */
  async listSkills(projectPath: string): Promise<Skill[]> {
    return scanSkills(projectPath);
  }
  async listAgents(projectPath: string): Promise<AgentConfig[]> {
    return scanAgents(projectPath);
  }

  createSession(options: SessionOptions): AdapterSession {
    // A history/resume read (getMessagesFromDisk) passes chatId. Reuse the matching cached session,
    // or fall back to the most recent live one — never advance the per-key index or load a new file.
    if (options.chatId) {
      const events = this.eventsBySessionId.get(options.chatId) ?? this.lastLiveEvents;
      if (events) return new ReplaySession(options, events);
    }

    const dir = process.env['E2E_RECORDINGS_DIR'];
    if (!dir) throw new Error('mock-cli requires E2E_RECORDINGS_DIR');
    const key = process.env['E2E_RECORDING_KEY'] ?? 'session';
    const index = this.indexByKey.get(key) ?? 0;
    this.indexByKey.set(key, index + 1);

    const file = join(dir, `${sanitizeKey(key)}.${index}.ndjson`);
    if (!existsSync(file)) {
      throw new Error(`mock-cli: fixture not found: ${file} — record it with E2E_MODE=record`);
    }
    const events = createReplayState(readFileSync(file, 'utf-8')).events;
    this.lastLiveEvents = events;

    // Cache by recorded sessionId (the onInit arg) for later history reads.
    const sessionId = events.find((e) => e.dir === 'out' && e.method === 'onInit')?.args[0];
    if (typeof sessionId === 'string') this.eventsBySessionId.set(sessionId, events);

    return new ReplaySession(options, events);
  }
}

File: packages/core-rs/crates/mainframe-adapter-mock/src/adapter.rs
use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};

use mainframe_adapter_api::{Adapter, AdapterError, AdapterSession, BoxFuture};
use mainframe_types::adapter::{AdapterCapabilities, AdapterModel, EffortLevel, SessionOptions};
use mainframe_types::display::ToolCategories;

use crate::session::{ReplayCache, ReplaySession};

#[derive(Default)]
pub struct MockCliAdapter {
    indexes: Mutex<HashMap<String, usize>>,
    cache: Arc<ReplayCache>,
}

pub fn sanitize_key(key: &str) -> String {
    let mut sanitized = String::new();
    let mut last_dash = false;
    for character in key.chars() {
        let valid = character.is_ascii_alphanumeric() || matches!(character, '.' | '_' | '-');
        let next = if valid { character } else { '-' };
        if next == '-' && last_dash {
            continue;
        }
        last_dash = next == '-';
        sanitized.push(next);
    }
    sanitized.trim_matches('-').to_string()
}

fn models() -> Vec<AdapterModel> {
    vec![
        model("claude-haiku-4-5-20251001", "Haiku 4.5", true, None),
        model(
            "claude-sonnet-4-5-20251101",
            "Sonnet 4.5",
            false,
            Some(vec![
                EffortLevel::Low,
                EffortLevel::Medium,
                EffortLevel::High,
                EffortLevel::Max,
            ]),
        ),
        model(
            "claude-opus-4-5-20251001",
            "Opus 4.5",
            false,
            Some(vec![
                EffortLevel::Low,
                EffortLevel::Medium,
                EffortLevel::High,
                EffortLevel::Xhigh,
                EffortLevel::Max,
            ]),
        ),
    ]
}

fn model(
    id: &str,
    label: &str,
    is_default: bool,
    efforts: Option<Vec<EffortLevel>>,
) -> AdapterModel {
    let capable = efforts.is_some();
    AdapterModel {
        id: id.to_string(),
        label: label.to_string(),
        description: None,
        resolved_model: None,
        context_window: None,
        is_default: is_default.then_some(true),
        supported_efforts: efforts,
        default_effort: capable.then_some(EffortLevel::Medium),
        supports_fast: capable.then_some(true),
        supports_ultracode: (id.contains("opus")).then_some(true),
        supports_adaptive_thinking: (id.contains("opus")).then_some(true),
        supports_personality: None,
    }
}

impl Adapter for MockCliAdapter {
    fn id(&self) -> &str {
        "mock-cli"
    }
    fn name(&self) -> &str {
        "Mock CLI"
    }
    fn capabilities(&self) -> AdapterCapabilities {
        AdapterCapabilities { plan_mode: true }
    }
    fn is_installed(&self) -> BoxFuture<'_, Result<bool, AdapterError>> {
        Box::pin(async { Ok(true) })
    }
    fn get_version(&self) -> BoxFuture<'_, Result<Option<String>, AdapterError>> {
        Box::pin(async { Ok(Some("0.1.0".to_string())) })
    }
    fn list_models(&self) -> BoxFuture<'_, Result<Vec<AdapterModel>, AdapterError>> {
        Box::pin(async { Ok(models()) })
    }
    fn get_fallback_models(&self) -> Option<Vec<AdapterModel>> {
        Some(models())
    }

    fn create_session(&self, options: SessionOptions) -> Arc<dyn AdapterSession> {
        if let Some(session_id) = options.chat_id.as_deref()
            && let Some(events) = self.cache.lookup(session_id)
        {
            return Arc::new(ReplaySession::new(options, events));
        }
        let recordings_dir = match std::env::var("E2E_RECORDINGS_DIR") {
            Ok(dir) => dir,
            Err(_) => {
                return Arc::new(ReplaySession::failed(
                    options,
                    "mock-cli requires E2E_RECORDINGS_DIR".to_string(),
                ));
            }
        };
        let key = std::env::var("E2E_RECORDING_KEY").unwrap_or_else(|_| "session".to_string());
        let index = {
            let mut indexes = self.indexes.lock().unwrap_or_else(|e| e.into_inner());
            let index = *indexes.get(&key).unwrap_or(&0);
            indexes.insert(key.clone(), index + 1);
            index
        };
        let path = std::path::Path::new(&recordings_dir)
            .join(format!("{}.{index}.ndjson", sanitize_key(&key)));
        Arc::new(ReplaySession::from_fixture(
            options,
            path,
            self.cache.clone(),
        ))
    }

    fn kill_all(&self) {}

    fn get_tool_categories(&self) -> Option<ToolCategories> {
        Some(ToolCategories {
            explore: HashSet::from_iter(["Read", "Glob", "Grep", "LS"].map(str::to_string)),
            hidden: HashSet::new(),
            progress: HashSet::from_iter(["TaskCreate", "TaskUpdate"].map(str::to_string)),
            subagent: HashSet::from_iter(["Task", "Agent"].map(str::to_string)),
        })
    }
}
